### PR TITLE
New resource 'azurerm_private_dns_mx_record'

### LINKED ISSUE #4914: Support for Private DNS MX records
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently the provider is missing a resource for Private DNS MX records.

### New or Affected Resource(s)

* azurerm_private_dns_mx_record

### Potential Terraform Configuration

```hcl
resource "azurerm_resource_group" "test" {
  name     = "prvdns-rg"
  location = "westeurope"
}

resource "azurerm_private_dns_zone" "test" {
  name                = "testzone.com"
  resource_group_name = "${azurerm_resource_group.test.name}"
}

resource "azurerm_private_dns_mx_record" "test" {
  resource_group_name = "${azurerm_resource_group.test.name}"
  zone_name           = "${azurerm_private_dns_zone.test.name}"
  ttl                 = 300
  record {
    preference = 10
    exchange   = "mx1.testzone.com"
  }

  record {
    preference = 10
    exchange   = "mx2.testzone.com"
  }
}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by phires**
Fixes #4914

### CODE REVIEW

File: azurerm/resource_arm_private_dns_mx_record.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+	return results
+}
+
+func expandAzureRmPrivateDnsMxRecords(d *schema.ResourceData) *[]privatedns.MxRecord {
+	recordStrings := d.Get("record").(*schema.Set).List()
+	records := make([]privatedns.MxRecord, len(recordStrings))
+
+	for i, v := range recordStrings {
+		record := v.(map[string]interface{})
```
Comment (mbfrahry) [inline azurerm/resource_arm_private_dns_mx_record.go]: `v` could be nil and we should confirm that it's not before casting it to a `map[string]interface{}`

File: azurerm/resource_arm_private_dns_mx_record_test.go [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMPrivateDnsMxRecordDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAzureRMPrivateDnsMxRecord_basic(ri, location),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMPrivateDnsMxRecordExists(resourceName),
+				),
+			},
```
Comment (mbfrahry) [inline azurerm/resource_arm_private_dns_mx_record_test.go]: We should do a sanity check that import works through each apply by adding the import test step between every test step.

Comment (phires) [reply azurerm/resource_arm_private_dns_mx_record_test.go]: Well, any idea on how to implement this?

Comment (mbfrahry) [reply azurerm/resource_arm_private_dns_mx_record_test.go]: Ah, my fault. I commented on the wrong line. We want
```
			{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateVerify: true,
			},
```
on lines:

https://github.com/terraform-providers/terraform-provider-azurerm/pull/4915/files#diff-e8c9d50e92f6adf13be4cb73347f85e9R117

https://github.com/terraform-providers/terraform-provider-azurerm/pull/4915/files#diff-e8c9d50e92f6adf13be4cb73347f85e9R94

and

https://github.com/terraform-providers/terraform-provider-azurerm/pull/4915/files#diff-e8c9d50e92f6adf13be4cb73347f85e9R87

File: azurerm/resource_arm_private_dns_mx_record.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	if err != nil {
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["MX"]
+	zoneName := id.Path["privateDnsZones"]
+
+	_, err = dnsClient.Delete(ctx, resGroup, zoneName, privatedns.MX, name, "")
+	if err != nil {
```
Comment (katbyte) [inline azurerm/resource_arm_private_dns_mx_record.go]: these lines could be combined:
```
	if _, err = dnsClient.Delete(ctx, resGroup, zoneName, privatedns.MX, name, ""); err != nil {
```

File: azurerm/resource_arm_private_dns_mx_record.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+	if records != nil {
+		for _, record := range *records {
+			if record.Preference == nil ||
+				record.Exchange == nil {
+				continue
+			}
+
+			results = append(results, map[string]interface{}{
+				"preference": *record.Preference,
+				"exchange":   *record.Exchange,
```
Comment (katbyte) [inline azurerm/resource_arm_private_dns_mx_record.go]: We'll need to nil check here

File: website/docs/r/private_dns_mx_record.html.markdown [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+## Example Usage
+
+```hcl
+resource "azurerm_resource_group" "example" {
+  name     = "acceptanceTestResourceGroup1"
+  location = "West US"
+}
+
+resource "azurerm_private_dns_zone" "test" {
```
Comment (katbyte) [inline website/docs/r/private_dns_mx_record.html.markdown]: Could we use example not test in docs?
```suggestion
resource "azurerm_private_dns_zone" "example" {
```

File: website/docs/r/private_dns_mx_record.html.markdown [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+The following attributes are exported:
+
+* `id` - The Private DNS MX Record ID.
+
+## Import
+
+Private DNS MX Records can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_private_dns_srv_record.test /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/mygroup1/providers/Microsoft.Network/privateDnsZones/contoso.com/MX/@
```
Comment (katbyte) [inline website/docs/r/private_dns_mx_record.html.markdown]:
```suggestion
terraform import azurerm_private_dns_srv_record.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/mygroup1/providers/Microsoft.Network/privateDnsZones/contoso.com/MX/@
```

File: azurerm/resource_arm_private_dns_mx_record_test.go [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+    exchange   = "mx2.contoso.com"
+  }
+}
+`, template)
+}
+
+func testAccAzureRMPrivateDnsMxRecord_updateRecords(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+	name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/resource_arm_private_dns_mx_record_test.go]:
```suggestion
	name     = "acctestRG-prvdns-%d"
```

File: azurerm/resource_arm_private_dns_mx_record_test.go [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+    exchange   = "backupmx.contoso.com"
+  }
+}
+`, rInt, location, rInt, rInt)
+}
+
+func testAccAzureRMPrivateDnsMxRecord_withTags(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+	name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/resource_arm_private_dns_mx_record_test.go]:
```suggestion
	name     = "acctestRG-prvdns-%d"
```

File: azurerm/resource_arm_private_dns_mx_record_test.go [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+    cost_center = "MSFT"
+  }
+}
+`, rInt, location, rInt, rInt)
+}
+
+func testAccAzureRMPrivateDnsMxRecord_withTagsUpdate(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+	name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/resource_arm_private_dns_mx_record_test.go]:
```suggestion
	name     = "acctestRG-prvdns-%d"
```

**Pull request merged**